optimizely/optimizely_config.py

```diff
@@ -0,0 +1,225 @@
+
+
+class OptimizelyFeature(object):
+    def __init__(self, id, key, experiments_map, variables_map):
+        self.id = id
+        self.key = key
+        self.experiments_map = experiments_map
+        self.variables_map = variables_map
+
+
+class OptimizelyVariation(object):
+    def __init__(self, id, key, feature_enabled, variables_map):
+        self.id = id
+        self.key = key
+        self.feature_enabled = feature_enabled
+        self.variables_map = variables_map
+
+
+class OptimizelyVariable(object):
+    def __init__(self, id, key, variable_type, value):
+        self.id = id
+        self.key = key
+        self.type = variable_type
+        self.value = value
+
+
+class OptimizelyConfigService(object):
+    """ Class encapsulating methods to be used in creating instance of OptimizelyConfig. """
+
+    def __init__(self, project_config):
+        """
+        Args:
+            project_config ProjectConfig
+        """
+        self.is_valid = True
+
+        if not isinstance(project_config, ProjectConfig):
+            self.is_valid = False
+            return
+
+        self.experiments = project_config.experiments
+        self.feature_flags = project_config.feature_flags
+        self.groups = project_config.groups
+        self.revision = project_config.revision
+
+        self._create_lookup_maps()
+
+    def get_config(self):
+        """ Gets instance of OptimizelyConfig
+
+        Returns:
+            Optimizely Config instance or None if OptimizelyConfigService is invalid.
+        """
+
+        if not self.is_valid:
+            return None
+
+        experiments_key_map, experiments_id_map = self._get_experiments_maps()
+        features_map = self._get_features_map(experiments_id_map)
+
+        return OptimizelyConfig(self.revision, experiments_key_map, features_map)
+
+    def _create_lookup_maps(self):
+        """ Creates lookup maps to avoid redundant iteration of config objects.  """
+
+        self.exp_id_to_feature_map = {}
+        self.feature_key_variable_key_to_variable_map = {}
+        self.feature_key_variable_id_to_variable_map = {}
+
+        for feature in self.feature_flags:
+            for experiment_id in feature['experimentIds']:
+                self.exp_id_to_feature_map[experiment_id] = feature
+
+            variables_key_map = {}
+            variables_id_map = {}
+            for variable in feature.get('variables', []):
+                opt_variable = OptimizelyVariable(
+                    variable['id'], variable['key'], variable['type'], variable['defaultValue']
+                )
+                variables_key_map[variable['key']] = opt_variable
+                variables_id_map[variable['id']] = opt_variable
+
+            self.feature_key_variable_key_to_variable_map[feature['key']] = variables_key_map
+            self.feature_key_variable_id_to_variable_map[feature['key']] = variables_id_map
+
+    def _get_variables_map(self, variation, experiment):
```

Comment:
nit. I personally prefer arranging this as `self, experiment, variation` to honor the parent child relationship between experiments and variations.